### PR TITLE
Deliver mails in context of their recipient

### PR DESCRIPTION
**Pull request by NobodysNightmare**
## OpenProject work package

https://community.openproject.org/work_packages/19769
## Description

This PR changes mail notification so that they set the current user to the recipient of the mail while it is delivered. Thereby it is ensured that methods like `WorkPackage.visible` work during mail creation. This is required for text formatting, e.g. to turn `##123` into `Bug 123: A bug!`

### CODE REVIEW

**Comment by TeatroIO** [comment]
I've prepared a stage to preview changes. <a href='http://nobodysnightmare-feature-recipient-context.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com' target='_blank'>Open stage</a> or <a href='http://nobodysnightmare-feature-recipient-context.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com/teatro.log' target='_blank'>view logs</a>.

File: app/workers/mail_notification_job.rb [hunk 1]
```diff
@@ -29,16 +29,20 @@
...
 # Requires including class to implement #notification_mail.
-module MailNotificationJob
+class MailNotificationJob
   mattr_accessor :raise_exceptions
 
-  def perform
-    notify
+  def initialize(recipient_id, author_id)
+    @recipient_id = recipient_id
+    @author_id    = author_id
```
Comment (NobodysNightmare) [inline app/workers/mail_notification_job.rb:37]: I do not agree and rubocop does not agree either.

**Comment by Azure7111** [comment]
So far this looks good. But the spec failures from the previous completed run seem legit. Lets wait what the next run of Travis finds. You may however add tests for `execute_as` in the meantime as that method has probably no coverage right now.

**Comment by ulferts** [comment]
@NobodysNightmare can you please add specs to this. At least the `execute_as` behaviour should be solidified by specs.

[edit] sorry, haven't read @Azure7111's comment before writing.

**Comment by ulferts** [comment]
@NobodysNightmare the specs refer to shared examples which do not exist.

**Comment by NobodysNightmare** [comment]
@ulferts Lets try again ^^

File: spec/workers/mail_notification_jobs/shared_examples.rb [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+
+      expect(mail).not_to be_present
+    end
+
+    it 'does not raise an error but fails silently' do
+      expect { job.perform }.not_to raise_error
+    end
+
+    context 'raising exceptions' do
+      before { MailNotificationJob.raise_exceptions = true }
```
Comment (myabc) [inline spec/workers/mail_notification_jobs/shared_examples.rb:70]: @NobodysNightmare try `around { |example| example.run }` instead

Comment (NobodysNightmare) [inline spec/workers/mail_notification_jobs/shared_examples.rb:70]: @myabc You are saying that:

``` ruby
before { MailNotificationJob.raise_exceptions = true }
after { MailNotificationJob.raise_exceptions = false }
```

and

``` ruby
around { |example|
  MailNotificationJob.raise_exceptions = true
  example.run
  MailNotificationJob.raise_exceptions = false
}
```

are equivalent?

This is nice to know, I'll try to remember the next time I am writing a spec of this kind. So thank you :)

However: I'd rather want this PR to be merged now. This code wouldn't be in the diff if it wasn't for the fact that I split up the specs for multiple classes into multiple files. My boyscout contribution for this spec is already done ;-)

Comment (myabc) [inline spec/workers/mail_notification_jobs/shared_examples.rb:70]: @NobodysNightmare yep, they're equivalent – it makes intention slightly clearer IMO.

**Pull request merged**